### ISSUE TITLE
Disable access to federated content for users that are not logged in

### ISSUE THREAD

**Issue by FireFury0** [mastodon/mastodon#22620]
### Pitch

Mastodon should allow administrators to disable access to federated toots (i.e. via the "Federated" timeline and search functions) when users are not logged in.

### Motivation

Although in many cases, admins can ensure that toots from their local users are appropriate, they only have limited control over the appropriateness of toots received from federated instances.  There may therefore be legal implications in republishing federated toots.

The administrator can place requirements on local users, such as ensuring they are adults, etc. so the risks associated with presenting federated toots to logged in users are limited.  However, users that are not logged in are currently allowed access to federated toots.  This opens up a whole host of legal problems, since the instance is republishing arbitrary content to unknown people.

Secondly, small instances may find it costly to allow people who aren't their users to access the federated timeline.

Even where anonymous access to federated content is undesirable, it is probably still desirable to allow anonymous access to local content.

**Comment by rbairwell** [comment]
Under Administration > Site Settings there is the "Include federated content on unauthenticated public timeline page" option (at least on my test instance).

**Comment by FireFury0** [comment]
> Under Administration > Site Settings there is the "Include federated content on unauthenticated public timeline page" option (at least on my test instance).

I don't have a "Site settings" - under Administration I have: Dashboard, Server Settings, Server rules, Roles, Announcements, Custom emojis, Webhooks, Relays.  I can't see "Include federated content on unauthenticated public timeline page" in any of those.

I'm on Mastodon 4.0.2.  Some Googling suggests that "Site settings" is from an older version, so maybe this feature used to exist and has gone away?  Grepping the source tree for "Include federated content" doesn't find anything.

It may have been removed by https://github.com/mastodon/mastodon/pull/19407  ?

**Comment by FireFury0** [comment]
Found it - it was removed by https://github.com/mastodon/mastodon/pull/19257

**Comment by nemobis** [comment]
Yes, the main source of the issue is that Mastodon 4 removed the logged-out UI (see also https://github.com/mastodon/mastodon/issues/20345 https://github.com/mastodon/mastodon/issues/20336). I guess that's not going to come back, but it should be relatively easy to provide

At the moment it's possible for logged-out users, by navigation from other pages, to load URLs like https://mamot.fr/@EU_Commission@social.network.europa.eu and https://mamot.fr/@EU_Commission@social.network.europa.eu/108209466785071868; Mastodon should offer an option to unconditionally open the target URLs they redirect to when opened in a new tab (in the example, https://social.network.europa.eu/@EU_Commission and https://social.network.europa.eu/@EU_Commission/108209466301247591). Cf. 
 https://github.com/mastodon/mastodon/issues/19902 for a case where we actually want the opposite.

Note that the Mastodon 4 UI also tries to catch all clicks and prevent the user from leaving the current domain (https://github.com/mastodon/mastodon/issues/20827 https://github.com/mastodon/mastodon/issues/22673) so there might be some tension with that.

In the meanwhile, you might be able to reduce the impact of the issue with some tweaks:
* enable [authorized_fetch](https://docs.joinmastodon.org/admin/config/#authorized_fetch);
* try to remove users and posts from other instances from the various recommendations on /search, /explore, /tags;
* make /search, /public (the federated timeline) and /explore impossible to load for logged-out users (maybe with some nginx setting?);
* make sure that / redirects to /about and not /explore.

https://eupolicy.social/ has implemented some but not all of these.

**Comment by brookmiles** [comment]
I would like this option as well, my concerns and motivation match the OP.

**Comment by ciencia** [comment]
Recently I've seen some controversy about content being republished (or relayed) into "nostr" and other networks, where users were concerned about losing control about their publications (their entire profile is cloned on other instances, like an impersonation). Some server admins simply blocked federation with the relay instances. However, this measure can be circumvented by retrieving users' timeline from another federated instance, where users on the other instance follow that account.

With the growing "ai" shitstorm of bots scraping content to build their model, this can be abused by bots, scraping content from other instances where they haven't been blocked or rate limited yet.

**Comment by nemobis** [comment]
In v4.3.3, I still find no way to prevent unauthenticated users from searching accounts on remote instances.

![Image](https://github.com/user-attachments/assets/d15d4c3e-13af-4d53-b6a1-b316f8010dfb)

**Comment by ciencia** [comment]
Actually, I think it has partially been fixed in recent versions. It's true you can see other profiles and publications when navigating through the browser (with JavaScript enabled), but if you try to open the link of an external profile or publication on a new tab/window, you get an interstitial page saying you're about to leave the current instance, and the full link is displayed (you have to click on it to navigate to the external instance).

At least, it protects instances against automated scraping though crawlers. I don't know if the same api that's using JavaScript to display the contents can be abused for this purpose, though.

**Comment by cooperaj** [comment]
https://github.com/mastodon/mastodon/discussions/28372

As I've commented on the above discussion being able to remove "Other Servers" from our public unauthenticated feeds would go a long way to reducing the risks we're going to have to assess under the UK's OAS legislation. 

I want people to access my stuff. I'm confident my stuff is negligible risk for everything, but as long as my mastodon is publishing a public feed that I have no control over those risks go *waaaay* up. Yes it's true that I can disable unauthenticated access in general but this then makes my own posts (and the replies to them) much less discoverable by 3rd party instance users which is a massive disadvantage.

**Comment by ClearlyClaire** [comment]
You can disable unauthenticated access to public timelines. I understand it does not fit everyone's use case, but for a single-user server, this should still work, as you would not disable unauthenticated access to your account or your posts.

![Image](https://github.com/user-attachments/assets/05427e30-096a-443b-8004-205fcef5214b)

**Comment by cooperaj** [comment]
> but for a single-user server, this should still work, as you would not disable unauthenticated access to your account or your posts.

Yes, thank you. I'd just had a play around and discovered that this was the case. Given the instance is in single user mode the main page (my profile) also happens to be the "local feed" so that's okay but If I were to be hosting a handful of users I'd probably want to retain the local feed as expected - so I still think this is a valid need.

**Comment by ThisIsMissEm** [comment]
I started work in https://github.com/mastodon/mastodon/pull/31471 however, wasn't happy with the overall approach and think a different approach is necessary

**Comment by ShadowJonathan** [comment]
Why is this issue closed? What PR/commit implements it?

**Comment by ThisIsMissEm** [comment]
@ShadowJonathan because it's functionally the same goal overall as https://github.com/mastodon/mastodon/issues/20831

It was closed as duplicate

**Comment by nemobis** [comment]
https://github.com/mastodon/mastodon/issues/20831 is a subset of what's being asked here. That issue is only about timelines, this is about all federated content including explore, trends, search.

**Comment by WKobes** [comment]
@andypiper please consider reopening this issue. As mentioned, the duplicated issue solved just a subset of this issue.

Most significant, it is not possible to hide federated posts/accounts in search results for unauthenticated users. E.g. the Dutch government has a pilot instance (https://social.overheid.nl/) with disabled registration. Users that want to interact can only do so from their instances. The search tool on the government instance itself ideally would not show any federated users, making sure visitors to the instance see governmental posts and users only.

**Comment by andypiper** [comment]
Thank you @WKobes - we are working on related features (see "Institution Support" on the [current roadmap](https://joinmastodon.org/roadmap)). Mastodon 4.5 already permits server owners to set a default landing page (#20831) and in the future the software will offer further customisation. We will the suggested search restriction to that area of work.
